cupy/linalg/einsum.py

```diff
@@ -0,0 +1,251 @@
+            # TODO(fukatani): add up at here if enable.
+            self.result = cupy.tensordot(self.result, ioperand, axes=0)
+        self.subscript = ''.join(self.subscripts)
+
+
+def get_dummy_labels(label_list):
+    dummy_label_set = set([])
+    count_dict = collections.Counter(label_list)
+    for label, count in count_dict.items():
+        if count >= 2:
+            dummy_label_set.add(label)
+    return dummy_label_set
+
+
+def einsum(*operands):
+    """Evaluates the Einstein summation convention on the operands.
+
+    Using the Einstein summation convention, many common multi-dimensional
+    array operations can be represented in a simple fashion.  This function
+    provides a way to compute such summations.
+
+    Args:
+        subscripts (str): Specifies the subscripts for summation.
+        operands (sequence of arrays): These are the arrays for the operation.
+
+    Returns:
+        cupy.ndarray:
+            The calculation based on the Einstein summation convention.
+
+    .. seealso:: :func:`numpy.einsum`
+    """
+
+    # TODO(fukatani): Support 'out', 'order', 'dtype', 'casting', 'optimize'
+
+    if not operands:
+        raise ValueError("must specify the einstein sum subscripts string and "
+                         "at least one operand, or at least one operand and "
+                         "its corresponding subscripts list")
+
+    subscripts = operands[0]
+    ioperands = operands[1:]
+
+    if not isinstance(subscripts, str):
+        raise TypeError("Current cupy einsum support only string subscripts")
+
+    # TODO(fukatani): Support '...'
+    if '.' in subscripts:
+        raise TypeError("Current cupy einsum does not support '...' ellipsis")
+
+    subscripts = subscripts.replace(' ', '')
+    irregular_chars = set(subscripts) - set(string.ascii_letters) - set('->,')
+    if irregular_chars:
+        pickup = list(irregular_chars)[0]
+        raise ValueError("invalid subscript '{}' in einstein sum subscripts "
+                         "string, subscripts must be letters".format(pickup))
+
+    converted_inputs = []
+    dtype = numpy.result_type(*ioperands)
+    for a in ioperands:
+        if isinstance(a, cupy.ndarray):
+            converted_inputs.append(a.astype(dtype))
+        else:
+            converted_inputs.append(cupy.asarray(a, dtype=dtype))
+
+    pos = subscripts.find('-')
+    if pos != -1 and subscripts[pos + 1] != '>':
+        raise ValueError("einstein sum subscript string does not contain "
+                         "proper '->' output specified")
+
+    arrow_pos = subscripts.find('->')
+    if arrow_pos == -1:
+        input_subscripts = subscripts
+        label_list = list(input_subscripts.replace(',', ''))
+        out_label_set = set(label_list) - get_dummy_labels(label_list)
+        output_subscript = ''.join(sorted(list(out_label_set)))
+    else:
+        input_subscripts = subscripts[:arrow_pos]
+        output_subscript = subscripts[arrow_pos + 2:]
```

Comment:
Does it correctly raise an error when ouput_subscript includes "," such as "ij->i,j"?